### ISSUE TITLE
JWT_AUTH_COOKIE makes APIs CSRF vulnerable.

### ISSUE THREAD

**Issue by mahmoodi** [GetBlimp/django-rest-framework-jwt#338]
Using the JWT_AUTH_COOKIE is vulnerable to CSRF attacks and this is not mentioned in the documentation nor mitigated against in the code.

**Comment by angvp** [comment]
How is that? please explain.

**Comment by Radobilly** [comment]
Maybe by brute forcing the default hs256 JWT_ALGORITHM as mentioned [here](https://auth0.com/blog/brute-forcing-hs256-is-possible-the-importance-of-using-strong-keys-to-sign-jwts/) and [here](https://github.com/lmammino/jwt-cracker)

**Comment by angvp** [comment]
@mahmoodi ? input please.

**Comment by klis87** [comment]
@mahmoodi is right. Because JWT is saved in a cookie, it will be automatically sent with every request, causing successful authentication without any need to add JWT to request header by the application. Because of this, it is generally vulterable to CSRF in exactly the same way as usual cookie session, so actually there is no need to provide any example here as general CSRF vulnarability exaples could be applied here.

**Comment by angvp** [comment]
Hey @klis87 thanks for jump in, yeah I got that from @Radobilly's links and some googling .. I will write some docs and will update this ticket so you can proofread and add/delete more info to the main docs.

**Comment by AdamStelmaszczyk** [comment]
Please correct me if I'm wrong, but seems to me that default Django CSRF protection, if used correctly, prevents problems:

https://docs.djangoproject.com/en/1.11/ref/csrf/

https://stormpath.com/blog/where-to-store-your-jwts-cookies-vs-html5-web-storage also says that:

> Leveraging your web app framework’s CSRF protection makes cookies rock solid for storing a JWT.

**Comment by EduardoNogueira** [comment]
Any update on this?

**Comment by bmpenuelas** [comment]
I noticed this risk and created a pull request adding an option to handle csrf protection when using JWT in http-only cookie. 

Now I see it's an open issue, for those interested, check out https://github.com/GetBlimp/django-rest-framework-jwt/pull/434 and maybe we can close this.

**Comment by CapedHero** [comment]
Perhaps this [post](https://stackoverflow.com/a/37396572/7483239) might be illuminating?

**Comment by bmpenuelas** [comment]
It is @CapedHero, that is the behaviour that you get when using #434.

**Comment by ajbeach2** [comment]
Enabling the django.middleware.csrf.CsrfViewMiddleware should mitigate csrf.  By default, I think anti csrf tokens are stored in a cookie you can access from javascript, and send in a  X-CSRFToken header.

see: https://docs.djangoproject.com/en/2.1/ref/csrf/#ajax

Just note: DjangoRestFramework disables CSRF protection for unauthenticated requests in the APIView class. You will need to manually setup csrf protection if you want those on unauthenticated endpoints (which would mostly be login/signup).

**Comment by paul110590** [comment]
As well as requiring CSRF protection, shouldn't JWT_AUTH_COOKIE also be set with the Secure flag?

**Comment by paul110590** [comment]
Perhaps use the Secure flag by default and have an option to turn it off then?

**Comment by pou426** [comment]
There is currently no way to customise the cookie settings. It would be really useful to have the choice to set them, especially the `secure` and `samesite` flags. Thanks!

**Comment by bmpenuelas** [comment]
Yes @pou426 you can customize it the way you want using CSRF_COOKIE_SECURE and CSRF_COOKIE_SAMESITE respectively.
